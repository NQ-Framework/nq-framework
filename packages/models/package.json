{
  "name": "@nqframework/models",
  "version": "4.2.0",
  "description": "Data Transfer Objects for the NQ Framework",
  "main": "build/index.js",
  "module": "build/index.es.js",
  "jsnext:main": "build/index.es.js",
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "start:dev": "tsc --watch",
    "test:watch": "jest --watch",
    "test:coverage": "jest --coverage",
    "format": "prettier --write \"src/**/*.ts\"",
    "lint": "tslint -p tsconfig.json",
    "prepublishOnly": "npm test && npm run lint && npm run build",
    "preversion": "npm run lint",
    "version": "npm run format && git add -A src",
    "postversion": "git push && git push --tags"
  },
  "files": [
    "build",
    "LICENCE"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/nq-framework/models"
  },
  "keywords": [],
  "author": "Milos Spasojevic",
  "license": "MIT",
  "devDependencies": {
    "@types/jest": "^25.2.3",
    "jest": "^25.5.4",
    "prettier": "^2.1.2",
    "rollup": "^2.33.1",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-typescript2": "^0.26.0",
    "ts-jest": "^25.5.1",
    "tslint": "^6.1.3",
    "tslint-config-prettier": "^1.18.0",
    "typescript": "^3.9.7"
  },
  "dependencies": {
    "@irreal/decorator-mapper": "^1.0.21",
    "reflect-metadata": "^0.1.13"
  }
}
