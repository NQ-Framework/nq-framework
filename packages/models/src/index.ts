export { Organization } from "./organization/organization";
export { OrganizationMember } from "./organization/organization-member";
export { DataSource } from "./data-fetching/data-source";
export { MsSqlConfiguration } from "./data-fetching/ms-sql-configuration";
export { ApiConfiguration } from "./data-fetching/api-configuration";
export { ScheduledJob } from "./scheduled-jobs/scheduled-job";
export { TriggeredJob } from "./triggered-jobs/triggered-job";
export { ConfigurationInterface } from "./jobs/configuration-interface";
export { LogConfiguration } from "./jobs/log-configuration";
export {
  StoredProcedureConfiguration,
  ProcedureParameter,
} from "./jobs/stored-procedure-configuration";
export { Action } from "./workflow/actions/action";
export { ActionInstance } from "./workflow/actions/action-instance";
export { ActionResult } from "./workflow/actions/action-result";
export { ActionHandler } from "./workflow/actions/action-handler";
export { ActionLink } from "./workflow/actions/action-link";
export { Workflow } from "./workflow/workflow";
export { WorkflowExecutionData } from "./workflow/workflow-execution-data";
export { WorkflowExecutionResult } from "./workflow/workflow-execution-result";
export { WorkflowExecutionContext } from "./workflow/workflow-execution-context";
