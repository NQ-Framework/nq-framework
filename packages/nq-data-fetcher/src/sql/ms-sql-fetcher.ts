import { MsSqlConfiguration } from "@nq-framework/models";
import { DataFetcherInterface } from "../dataFetcherInterface";
import { Request, Connection, TediousType } from "tedious"

export class MsSqlFetcher implements DataFetcherInterface {
    constructor(private config: MsSqlConfiguration) { }
    async get(props: MsSqlFetcherProps): Promise<MsSqlDataResult> {
        const connection = await connect(this.config);
        return await execute(props, connection);
    }
}

export interface MsSqlFetcherProps {
    query: string
    isProcedure: boolean
    params?: MsSqlParameterType[]
}
export type MsSqlParameterType = {
    name: string,
    value: any,
    type: TediousType
}

export type MsSqlDataResult = {
    rowCount: number,
    rows: any[]
}

async function execute(props: MsSqlFetcherProps, connection: Connection): Promise<MsSqlDataResult> {
    return new Promise((resolve, reject) => {

        const request = new Request(props.query, (err, rowCount, rows) => {
            console.log('omg ', err, rowCount, rows);
            if (err) {
                reject(err);
                return;
            }
            resolve({ rowCount, rows })
        });
        if (props.params) {
            props.params.forEach(p => { request.addParameter(p.name, p.type, p.value) });
        }
        if (props.isProcedure) {
            connection.callProcedure(request);
        }
        else {
            connection.execSql(request);
        }
    });


}

async function connect(config: MsSqlConfiguration): Promise<Connection> {
    return new Promise((resolve, reject) => {
        var connection = new Connection({
            server: config.serverIp,
            authentication: {
                options: {
                    userName: config.username,
                    password: config.password,
                },
                type: 'default'
            },
            options: {
                trustServerCertificate: config.trustServerCertificate,
                database: config.database,
                rowCollectionOnRequestCompletion: true
            }
        });
        connection.connect((err) => {
            if (!err) {
                resolve(connection);
                return;
            }
            reject(err);
        });
    });
}