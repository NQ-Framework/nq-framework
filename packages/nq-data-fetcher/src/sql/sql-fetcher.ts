import { MsSqlConfiguration } from "@nq-framework/models";
import { DataFetcherInterface } from "../dataFetcherInterface";

export class SqlFetcher implements DataFetcherInterface {
    constructor(private config: MsSqlConfiguration) { }
    async get(props: any): Promise<{ data: any; }> {
        console.log('props ', props);
        return { data: "this is sample data obtained using: " + JSON.stringify(this.config) };
    }

}