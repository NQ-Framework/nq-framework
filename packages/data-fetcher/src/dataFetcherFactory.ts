import { DataSource } from "@nqframework/models";
import { MsSqlConfiguration } from "@nqframework/models";
import { DataFetcherInterface } from "./dataFetcherInterface";
import { MsSqlFetcher } from "./sql/ms-sql-fetcher";

export class DataFetcherFactory {
    create(dataSource: DataSource): DataFetcherInterface {
        if (dataSource.configuration.type === "mssql") {
            return new MsSqlFetcher(dataSource.configuration as MsSqlConfiguration);
        }
        throw new Error("Could not create Fetcher for request " + JSON.stringify(dataSource));
    }
}