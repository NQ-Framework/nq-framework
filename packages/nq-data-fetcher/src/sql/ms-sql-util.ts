import { MsSqlConfiguration } from "@nqframework/models";
import { Connection, Request } from "tedious";
import { MsSqlFetcherProps, MsSqlDataResult } from "./ms-sql-fetcher";

export async function execute(props: MsSqlFetcherProps, connection: Connection): Promise<MsSqlDataResult> {
    return new Promise((resolve, reject) => {

        const rows: any[] = [];
        const outParams: any = {};
        const request = new Request(props.query, (err) => {
            connection.close();

            if (err) {
                reject(err);
                return;
            }
            resolve({ rowCount: rows.length, rows, outParams })
        });
        request.on("row", (columns) => {
            rows.push(columns.filter(c => c.metadata.colName !== 'metadata').reduce((row, c) => {
                row[c.metadata.colName] = c.value;
                return row;
            }, {}));
        });
        request.on("returnValue", (name, value) => {
            outParams[name] = value;
        })
        if (props.params) {
            props.params.forEach(p => { request.addParameter(p.name, p.type, p.value) });
        }
        if (props.outParams) {
            props.outParams.forEach(p => { request.addOutputParameter(p.name, p.type) });
        }
        if (props.isProcedure) {
            connection.callProcedure(request);
        }
        else {
            connection.execSql(request);
        }
    });
}

export async function connect(config: MsSqlConfiguration): Promise<Connection> {
    return new Promise((resolve, reject) => {
        let ipAddress = config.serverIp;
        let port = '1433';
        if (ipAddress.includes(',')) {
            port = ipAddress.slice(ipAddress.indexOf(',') + 1);
            ipAddress = ipAddress.slice(0, ipAddress.indexOf(','));
        }
        if (config.port) {
            port = config.port;
        }
        var connection = new Connection({
            server: ipAddress,
            authentication: {
                options: {
                    userName: config.username,
                    password: config.password,
                },
                type: 'default'
            },
            options: {
                trustServerCertificate: config.trustServerCertificate,
                database: config.database,
                port: Number.parseInt(port)
            }
        });
        connection.connect((err) => {
            if (!err) {
                resolve(connection);
                return;
            }
            reject(err);
        });
    });
}