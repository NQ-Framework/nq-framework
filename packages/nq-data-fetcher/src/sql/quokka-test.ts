import { MsSqlFetcher } from "./ms-sql-fetcher";
import { TYPES } from "tedious";
const fetcher = new MsSqlFetcher({
    serverIp: 'localhost',
    username: 'sa',
    password: 'Test.1234!!',
    trustServerCertificate: true,
    database: 'TestDB-SM',
    type: 'mssql'
});

fetcher.get({
    isProcedure: true,
    query: 'testProc',
    params: [{
        name: "TestParam",
        type: TYPES.Int,
        value: 3
    }],
    outParams: [
        {
            name: "TestOut",
            type: TYPES.Int
        }
    ]
}).then(data => {

    console.log(data);
}).catch(err => {
    console.error('erro je', err);
})