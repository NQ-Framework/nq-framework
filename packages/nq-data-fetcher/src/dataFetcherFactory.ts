import { DataSource } from "@nq-framework/models";
import { MsSqlConfiguration } from "@nq-framework/models";
import { DataFetcherInterface } from "./dataFetcherInterface";
import { SqlFetcher } from "./sql/sql-fetcher";

export class DataFetcherFactory {
    create(dataSource: DataSource): DataFetcherInterface {
        if (dataSource.configuration.type === "mssql") {
            return new SqlFetcher(dataSource.configuration as MsSqlConfiguration);
        }
        throw new Error("Could not create Fetcher for request " + JSON.stringify(dataSource));
    }
}