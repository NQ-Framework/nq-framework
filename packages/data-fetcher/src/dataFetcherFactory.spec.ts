import { DataCredentials, reducePropertyValuesToObject } from "@nqframework/models";
import { DataFetcherFactory } from "./dataFetcherFactory"
import { MsSqlFetcher } from "./sql/ms-sql-fetcher"
jest.mock('./sql/ms-sql-fetcher');
const mockCredentials: DataCredentials = {
    name: 'mock credentials',
    isRemote: false,
    configuration: [
        { name: 'serverIp', value: "test" },
        { name: 'port', value: "test" },
        { name: 'username', value: "test" },
        { name: 'password', value: "test" },
        { name: 'database', value: "test" },
        { name: 'type', value: "mssql" },
        { name: 'trustServerCertificate', value: true },
    ],
    credentialsType: {
        description: 'mock desc',
        isEnabled: true,
        name: 'mock credentials type',
        properties: [],
        type: 'mssql'
    }
}
describe('DataFetcherFactory', () => {
    it('should be defined', () => {
        expect(new DataFetcherFactory()).toBeDefined();
    });
    it('should return sql fetcher if config type is sql', () => {
        const fetcher = new DataFetcherFactory().create("mssql", mockCredentials.configuration)
        expect(fetcher).toBeDefined();
        const mockFetcher = MsSqlFetcher as jest.Mock;
        expect(mockFetcher.mock.instances.length).toEqual(1);
        expect(mockFetcher).toHaveBeenCalledWith(reducePropertyValuesToObject(mockCredentials.configuration));
    });

    it('should throw for invalid configurations', () => {
        expect(() => {
            new DataFetcherFactory().create("invalid", [])
        }).toThrowErrorMatchingSnapshot();
    });
});
