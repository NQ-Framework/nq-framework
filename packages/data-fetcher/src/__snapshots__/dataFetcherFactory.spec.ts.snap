// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataFetcherFactory should throw for invalid configurations 1`] = `"Could not create Fetcher for request {\\"configuration\\":{\\"type\\":\\"non-sql\\"}}"`;
